### ISSUE TITLE
Plagued by ORA-08177: can't serialize access for this transaction

### ISSUE THREAD

**Issue by DKroot** [laurenz/oracle_fdw#250]
We use `oracle_fdw` for bulk ETL between Postgres and Oracle and while updates on Postgres side work very well, updates on Oracle side are plagued by random but frequent `ORA-08177: can't serialize access for this transaction`.

I went through all the docs and past issues on the topic and tried the following:
1. Using `CREATE TABLE ... SEGMENT CREATION IMMEDIATE`
1. Using `TRUNCATE TABLE ... REUSE STORAGE`
1. Using retry logic
1. Removing all indexes by using staging Oracle tables
1. Removing any concurrent access whatsoever by using staging Oracle tables

None of these worked. What else can I try to resolve ths issue?

The test case that is currently randomly failing is the following:
1. Oracle 12c
```SQL
CREATE TABLE stg_cg_cited_reference_gen2 (
  silverchair_id VARCHAR2(30)              NOT NULL ENABLE,
  gen1_wos_uid   VARCHAR2(30) DEFAULT NULL NOT NULL ENABLE,
  gen1_pmid      NUMBER,
  gen2_wos_uid   VARCHAR2(30),
  gen2_pmid      NUMBER,
  gen2_pmcid     VARCHAR2(30),
  source_title   VARCHAR2(300),
  issue          VARCHAR2(10),
  volume         VARCHAR2(20),
  document_title VARCHAR2(2000),
  full_name      CLOB
) SEGMENT CREATION IMMEDIATE PCTFREE 10 PCTUSED 40 INITRANS 1 MAXTRANS 255 NOCOMPRESS LOGGING STORAGE (INITIAL 81920 NEXT 1048576 MINEXTENTS 1 MAXEXTENTS 2147483645 PCTINCREASE 0 FREELISTS 1 FREELIST GROUPS 1 BUFFER_POOL DEFAULT FLASH_CACHE DEFAULT CELL_FLASH_CACHE DEFAULT
) TABLESPACE users LOB (full_name) STORE AS SECUREFILE (TABLESPACE users ENABLE STORAGE IN ROW CHUNK 16384 NOCACHE LOGGING NOCOMPRESS KEEP_DUPLICATES STORAGE (INITIAL 212992 NEXT 1048576 MINEXTENTS 1 MAXEXTENTS 2147483645 PCTINCREASE 0 BUFFER_POOL DEFAULT FLASH_CACHE DEFAULT CELL_FLASH_CACHE DEFAULT
)
)
/
```

or if the table already exists:

```SQL
TRUNCATE TABLE stg_cg_cited_reference_gen2 REUSE STORAGE;
```

2. Postgres 9.6.8
```SQL
CREATE TABLE temp_cg_g2 (
  silverchair_id VARCHAR(30),
  gen1_wos_uid   VARCHAR(30),
  gen1_pmid      INTEGER,
  gen2_wos_uid   VARCHAR(30),
  gen2_pmid      INTEGER,
  gen2_pmcid     VARCHAR,
  source_title   VARCHAR(300),
  issue          VARCHAR(10),
  volume         VARCHAR(20),
  document_title VARCHAR(2000),
  full_name      TEXT
);
```

3. Populate data and execute:
```
INSERT INTO foreign_irdb.stg_cg_cited_reference_gen2
   SELECT *
   FROM temp_cg_g2;
```

The query fails randomly after e.g. about 50 minutes.

Environment:
`SELECT oracle_diag('irdb');`:

`oracle_fdw 2.0.0, PostgreSQL 9.6.8, Oracle client 12.2.0.1.0, Oracle server 12.1.0.2.0`

**Comment by laurenz** [comment]
It shouldn't be the problem from #200, because there are no indexes, and deferred segmeent creation can only strike during the first insert.

I have no idea, but my feeling that Oracle really goofed up when they implemented `SERIALIZABLE` transactions is becoming stronger and stronger.

There is no concurrent write access? You are inserting in a single thread?

**Comment by DKroot** [comment]
Yes, there should absolutely zero concurrent access: my last attempt at a workaround was using staging tables (that I referenced in the issue). These staging tables are not accessible at all to other users and are only used for one INSERT by my user account at a time. The only possibility of concurrent access that I can think of is that if DBAs of this enterprise DB are accessing these tables for some sort of system queries.

Is it possible for you to decrease isolation level if I use, e.g. `BEGIN READ COMMITTED` in Postgres SQL?

**Comment by DKroot** [comment]
Would it make sense to try to partition the Oracle table?

**Comment by laurenz** [comment]
This is a frequently asked question in that context, but no, oracle_fdw needs `SERIALIZABLE` transactions. It is easily possible that one PostgreSQL query causes several Oracle queries (inner side of a nested loop join, several foreign tables involved in one query, ...), and these queries need to see a consistent state of the database.

There are no indexes on the table, you say. Can you verify with

    SELECT index_name FROM user_indexes
       WHERE table_owner = 'SCHEMANAME' AND table_name = 'TABLENAME';

I really have no clue what could be the problem other than concurrent modifications in combination with indexes.

You could ask Oracle, but they might just tell you that serialization errors are to be expected...

You can change the oracle_fdw source [here](https://github.com/laurenz/oracle_fdw/blob/master/oracle_utils.c#L477) and replace `OCI_TRANS_SERIALIZABLE` with `0`. Then it should use `READ COMMITTED` and the errors will vanish.

That should be OK as long as all you do are simple inserts. But be warned that the result of more complicated queries might be wrong in the face of concurrent database activity.

**Comment by DKroot** [comment]
Verified: there are no indexes.

I'll try to recompile from the source per your suggestion. We do simple inserts for bulk ETL only on our project, so this workaround should work for us quite well.

Please consider providing some sort of option for this, so that people with similar problems can switch to `READ COMMITTED` more easily.

**Comment by silvioq** [comment]
> I verified that replacing OCI_TRANS_SERIALIZABLE with 0 fixes the issue. For those who need it, the fork is available [here](https://github.com/NETESOLUTIONS/oracle_fdw_read_committed).

This patch works like charm in my configuration. https://github.com/NETESOLUTIONS/oracle_fdw_read_committed/commit/f12acd8bc1ef151e685397a03c5587c55fe42132


```
# SELECT oracle_diag('orasircreb');
                                     oracle_diag                                      
--------------------------------------------------------------------------------------
 oracle_fdw 2.1.0, PostgreSQL 10.4, Oracle client 0.0.0.0.0, Oracle server 10.1.0.3.0

```

**Comment by laurenz** [comment]
Closed by #383.
